Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/checkins/CheckInsFragment.kt

```diff
@@ -57,33 +67,77 @@ class CheckInsFragment : Fragment(R.layout.trace_location_attendee_checkins_frag
-            addItemDecoration(TopBottomPaddingDecorator(topPadding = R.dimen.spacing_tiny))
-            itemAnimator = DefaultItemAnimator()
-            onScroll { extend ->
-                if (extend) binding.scanCheckinQrcodeFab.extend() else binding.scanCheckinQrcodeFab.shrink()
-            }
+        bindRecycler()
+        bindFAB()
 
-            onSwipeItem(
-                context = requireContext(),
-                excludedPositions = listOf() // TODO exclude items from swiping such as Camera permission item
-            ) { position, direction ->
-                val checkInsItem = checkInsAdapter.data[position]
-                if (checkInsItem.isSwipeable()) {
-                    checkInsItem.onSwipe(position, direction)
-                }
-            }
+        viewModel.checkins.observe2(this) { items ->
+            updateViews(items)
         }
 
-        viewModel.checkins.observe2(this) {
-            checkInsAdapter.update(it)
-            binding.apply {
-                checkInsList.isGone = it.isEmpty()
-                emptyListInfoContainer.isGone = it.isNotEmpty()
-            }
+        viewModel.events.observe2(this) {
+            onNavigationEvent(it)
+        }
+
+        viewModel.errorEvent.observe2(this) {
+            val errorForHumans = it.tryHumanReadableError(requireContext())
+            Toast.makeText(requireContext(), errorForHumans.description, Toast.LENGTH_LONG).show()
+        }
+    }
+
+    private fun onNavigationEvent(event: CheckInEvent?) {
+        when (event) {
+            is CheckInEvent.ConfirmCheckIn -> doNavigate(
+                CheckInsFragmentDirections.actionCheckInsFragmentToConfirmCheckInFragment(
+                    verifiedTraceLocation = event.verifiedTraceLocation,
+                    editCheckInId = 0,
+                )
+            )
+
+            is CheckInEvent.ConfirmSwipeItem -> showRemovalConfirmation(event.checkIn, event.position)
+
+            is CheckInEvent.ConfirmRemoveItem -> showRemovalConfirmation(event.checkIn, null)
+
+            is CheckInEvent.ConfirmRemoveAll -> showRemovalConfirmation(null, null)
+
+            is CheckInEvent.EditCheckIn -> doNavigate(
+                CheckInsFragmentDirections.actionCheckInsFragmentToConfirmCheckInFragment(
+                    verifiedTraceLocation = null,
+                    editCheckInId = event.checkInId,
+                )
+            )
+
+            is CheckInEvent.ShowInformation -> Toast.makeText(
+                requireContext(),
+                "TODO ¯\\_(ツ)_/¯",
+                Toast.LENGTH_SHORT
+            ).show()
+
+            is CheckInEvent.OpenDeviceSettings -> openDeviceSettings()
+        }
+    }
+
+    private fun updateViews(items: List<CheckInsItem>) {
+        checkInsAdapter.update(items)
+        binding.apply {
+            val index = items.indexOfFirst { it is CameraPermissionVH.Item }
+            swipeExcludedPositions.replaceAll(listOf(index))
```

Comment:
This doesn't feel very modular :frowning:.
Can't we do something like `SwipeExtension.getMovementFlags(...) { viewHolder isSwipable }`?

Reply:
Why do you think it is not modular. Is not the adapter itself doing the same? I don't like that we have to implement many interfaces for the data item and the holders just for that . Most of the screens does have to exclude swiping items , your suggestion will require more work for evey consumer to implement those interfaces one for the item which is OK and currently happening and one more for the holder.

Reply:
This isn't a blocker for me. Up to you.
I prefer list items to be self-sufficient and not requiring top-levels item specific logic.